README.md

```diff
@@ -1,16 +1,1261 @@
+
+A Promise is a placeholder value MAY be used as a variable placeholder for a concrete value in a [Closure](#3-closure), waiting on a previous step to complete.
+
+One way of seeing the names in a [`Batch`](#5-batch) is as variables for the return of each Closure. These can now be referenced by other Closures.
+
+For example, consider the following batch:
+
+``` json
+{
+  "run": {
+    "create-draft": {
+      "with": "https://example.com/blog/posts",
+      "do": "crud/create",
+      "inputs": {
+        "payload": {
+          "title": "How UCAN Tasks Changed My Life",
+          "body": "This is the story of how one spec changed everything..."
+        }
+      }
+    },
+    "get-editors": {
+      "with": "https://example.com/users/editors",
+      "do": "crud/read"
+    },
+    "notify": {
+      "with": "mailto:akiko@example.com",
+      "do": "msg/send",
+      "inputs": {
+        "to": {"ucan/ok": ["/", "right"]},
+        "subject": "Coffee",
+        "body": {"ucan/ok": ["/", "left"]}
+      }
+    }
+  }
+}
+```
+
+By analogy, this can be interpreted roughly as follows:
+
+``` js
+const createDraft = crud.create("https://example.com/blog/posts", {
+  payload: {
+    title: "How UCAN Tasks Changed My Life",
+    body: "This is the story of how one spec changed everything..."
+  }
+})
+
+const getEditors = crud.read("https://example.com/users/editors")
+
+const notify = msg.send("mailto:akiko@example.com", {
+  "to": await createDraft,
+  "subject": "Coffee",
+  "body": await getEditors
+})
+```
+
+While a Promise MAY be substituted for any field in a Closure, substituting the `do` field is NOT RECOMMENDED. The `do` field is critical in understanding what kind of action will be performed, and most schedulers will use the 
+
+After resolution, the Closure MUST be validated against the UCANs known to the Executor. A Promise resolved to a Closure that is not backed by a valid UCAN MUST NOT be executed, and SHOULD return an unauthorized error to the user.
+
+Promises MAY be used inside of a single Invocation's Closures, or across multiple Invocations, and MAY even be across multiple Invokers. As long as the pointer can be resolved, any invoked Task MAY be promised. This is sometimes referred to as ["promise pipelining"](http://erights.org/elib/distrib/pipeline.html).
+
+A Promise MUST resolve to a [Result](#8-result). If a particular branch's value is required to be unwrapped, the Result tag (`ok` or `err`) MAY be supplied.
+
+## 10.1 Fields
+
+``` ipldsch
+type Promise union {
+  | InvokedTaskPointer "ucan/ok"
+  | InvokedTaskPointer "ucan/err"
+  | InvokedTaskPointer "ucan/ptr"
```

Comment:
Is this meant to represent not yet unwrapped promise ? Might be worth making a note.

Reply:
I also kind of like sticking to `*` convention we have in UCANs so `ucan/*` to imply either. Alternatively `ucan/_` could mean either one.

Reply:
I can see us wanting to use `ucan/*` in lots of places in the future, so I'd like to at least disambiguate the namespace. We could also use `promise/*`, `promise/ok`, and `promise/err` if that's better.

Reply:
@Gozala updated!